### PR TITLE
[5.2] Fix key collisions when adding closures to global scopes

### PR DESCRIPTION
**Pull request by vlakoff**
Fix for #12530.

Unless @acasar can think of a better solution :)

### CODE REVIEW

**Comment by acasar** [comment]
@vlakoff Thanks for taking a look into this. :+1: 

I was maybe thinking to remove all random stuff an let php choose the next free index:

```
return static::$globalScopes[get_called_class()][] = $scope;
```

What do you think?

**Comment by vlakoff** [comment]
I thought about this too. But I feared collision with some named scope `0`.

A maybe better implementation would be a stack, not only for closures but also named scopes. It would still require a map, allowing user to get a registered scope by its name.

I thought about this so the scopes would be executed in the order they have been registered.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -352,7 +352,7 @@ public static function addGlobalScope($scope, Closure $implementation = null)
         }
 
         if ($scope instanceof Closure) {
-            return static::$globalScopes[get_called_class()][uniqid('scope')] = $scope;
+            return static::$globalScopes[get_called_class()][Str::random(32)] = $scope;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: Why this and not just random_bytes?

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: Wait, doesn't php have a way to get the id of a closure?

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: https://3v4l.org/C9T6k

Comment (vlakoff) [inline src/Illuminate/Database/Eloquent/Model.php]: Seems pretty good :smiley:

**Comment by vlakoff** [comment]
Updated PR using the great suggestion of @GrahamCampbell.

**Comment by acasar** [comment]
:+1:

**Pull request merged**